### PR TITLE
Mp 136 add tagging stsyem

### PR DESCRIPTION
**Pull request by BsBou**
- Created tags table
- Created taggings join table, which joins tags to pathways, challenges and profiles (the same tag can be associated to different models)
- Pathways are associated to tags on seed
- Tags can be associated to pathways on creation and edit
- Tags can be associated to challenges on creation and edit

### CODE REVIEW

File: app/views/pathways/_challenge.html.erb [hunk 1]
```diff
@@ -2,6 +2,9 @@
 <% challenge_template = capture do %>
    <div class='m-6 border p-4 flex flex-col hover:border-blue-500 w-full'>
       <h3 class='font-bold'><%= challenge.title %></h3>
+      <p><% challenge.tags.each do |tag| %>
+         (<%= tag.name %>)
```
Comment (adrianHards) [inline app/views/pathways/_challenge.html.erb:6]: if you could add a tag class, we can style it in sass later

File: app/controllers/challenges_controller.rb [hunk 2]
```diff
@@ -15,6 +15,7 @@ def create
     @challenge = Challenge.new(challenge_params)
     @challenge.user = @user
     if @challenge.save
+      add_tags_on_challenge_create(@challenge)
```
Comment (adrianHards) [inline app/controllers/challenges_controller.rb:18]: does this break if there are no tags?

File: db/seeds.rb [hunk 3]
```diff
@@ -15,11 +16,11 @@
 )
 
 User.create!(
-  first_name: "Mentor", last_name: "Fresh", email: "mentor@fresh.com", password: "Testpassword1!", role: "mentor"
+   first_name: "Mentor", last_name: "Fresh", email: "mentor@fresh.com", password: "Testpassword1!", role: "mentor"
 )
 
 User.create!(
-  first_name: "Mentee", last_name: "Fresh", email: "mentee@fresh.com", password: "Testpassword1!", role: "mentee"
+   first_name: "Mentee", last_name: "Fresh", email: "mentee@fresh.com", password: "Testpassword1!", role: "mentee"
```
Comment (adrianHards) [inline db/seeds.rb:23]: 🤔

File: spec/models/tag_spec.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+RSpec.describe Tag, type: :model do
+   describe 'Validations' do
+      it 'is valid' do
+         tag = build(:tag)
+         expect(tag).to be_an_instance_of Tag
+      end
+      
+      it 'has a name as a string' do
+         tag = create(:tag, name: 'Python')
```
Comment (adrianHards) [inline spec/models/tag_spec.rb:11]: prefer build() because create() persists

File: app/controllers/pathways_controller.rb [hunk 5]
```diff
@@ -18,21 +18,52 @@ def new
...
 
   def update
-    @pathway.update(pathway_params)
-    redirect_to pathway_path(@pathway)
+   @pathway.update(pathway_params)
+   edit_pathway_tags(@pathway)
+   redirect_to pathway_path(@pathway)
+  end
+
+  def edit_pathway_tags(pathway)
```
Comment (adrianHards) [inline app/controllers/pathways_controller.rb:47]: these methods should probably be in private (convention is for public methods to be route actions I believe).

**Review by adrianHards — CHANGES_REQUESTED** [review]
not sure we should allow users to add their own tags; controlling what they can select would be better imo

File: app/controllers/challenges_controller.rb (hunk 2 above)
Comment (BsBou) [reply app/controllers/challenges_controller.rb:18]: Still works!

File: app/views/pathways/_challenge.html.erb (hunk 1 above)
Comment (adrianHards) [reply app/views/pathways/_challenge.html.erb:6]: I just mean something like `:class => 'class'`, won't add any styling but will make it easier for those who do later on. Can always add later though.

File: app/controllers/pathways_controller.rb (hunk 5 above)
Comment (BsBou) [reply app/controllers/pathways_controller.rb:54]: all done!

**Comment by CraigNorford** [comment]
> 

Make a note of this question it's a good point to bring up in the next tech design. @BsBou also made a very good point in his video. I have some ideas

**Pull request merged**